pkg/terraform/exec/plugins/plugin_versions_generate.go

```diff
@@ -0,0 +1,249 @@
+			}
+		}
+	}
+}
+
+// getTerraformPluginVersions gathers the terraform provider version numbers
+func getTerraformPluginVersions(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+
+	// Output regex - clean up git name-rev output
+	ore := regexp.MustCompile(`[[:space:]]+|\^[[:digit:]]+`)
+
+	// Tag regex - match the git tag and hash
+	tre := regexp.MustCompile(`[[:digit:]]{14}-([a-fA-F0-9]{12})$`)
+
+	// URL regex - chop off anything after the terraform provider name
+	ure := regexp.MustCompile(`^(.+terraform-provider-[[:alnum:]]+)/v[[:digit:]]+`)
+
+	// Version regex - chop off anything that is not the version
+	vre := regexp.MustCompile(`([0-9\.]+)`)
+
+	for p, pi := range *plugins {
+		var repo *git.Repository
+		var mod *module.Version = &pi.Module
+
+		// XXX: we assume https checkouts
+		url := fmt.Sprintf("https://%s", mod.Path)
+
+		// If this is a v2 or up  module, chop off the end
+		matches := ure.FindStringSubmatch(mod.Path)
+		if len(matches) == 2 {
+			url = fmt.Sprintf("https://%s", matches[1])
+		}
+
+		// git clone if directory doesn't exist, else git pull
+		tmpDir := filepath.Join("tmp", p)
+		if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
+			os.MkdirAll(tmpDir, 0700)
+			repo, err = git.PlainClone(tmpDir, false, &git.CloneOptions{
+				URL:      url,
+				Progress: os.Stdout,
+			})
+			if err != nil {
+				log.Fatal("failed to clone repo ", url)
+			}
+		} else {
+			repo, err = git.PlainOpen(tmpDir)
+			if err != nil {
+				log.Fatal("failed to open repo ", tmpDir)
+			}
+			worktree, err := repo.Worktree()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = worktree.Pull(&git.PullOptions{})
+			if err != nil && err != git.NoErrAlreadyUpToDate {
+				log.Fatal(err)
+			}
+		}
+
+		pi.Tag = mod.Version
+
+		// If the version is a pseudo version, get the commit hash and
+		// resolve it to a tag if one exists, else default to v1.0.0
+		matches = tre.FindStringSubmatch(mod.Version)
+		if len(matches) == 2 {
+			h := matches[1]
+			hash, err := repo.ResolveRevision(plumbing.Revision(h))
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			here, err := os.Getwd()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = os.Chdir(tmpDir)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// XXX: HACK: fork git process until go-git supports name-rev
+			out, err := exec.Command("git", "name-rev", "--tags", "--name-only", hash.String()).Output()
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			// XXX:  What should be the default here?
+			tag := ore.ReplaceAllString(string(out), "")
+			if tag == "undefined" {
+				tag = "v1.0.0"
+			}
+			pi.Tag = tag
+
+			err = os.Chdir(here)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+
+		// Extract version out of tag
+		pi.Version = vre.FindString(pi.Tag)
+	}
+}
+
+// getLocalTerraformPluginVersions gathers the terraform provider version
+// numbers for local plugins
+func getLocalTerraformPluginVersions(file *modfile.File, plugins *map[string]*TFPluginInfo) {
```

Comment:
I have updated these functions to take and return a map and not use a pointer. Just out of curiosity, why is this better form than how I did it? The functions were manipulating the plugins variable. If this was C, this would be good form ;-) I am still new to golang, so this has go me curious why it's not good to do this.

Reply:
My main reasoning is that it is uncommon to see dereferences of a map, so it makes me have to think about what is going on. There is likely little performance difference between the two, so I would prefer to go with the one that is easy to read and rationalize about. For me it is easier to rationalize about a read-only input parameter and an output parameter, similar to `append`.

One major difference between C and GO is that C has a deterministic memory model. In Go, we have to deal with garbage collection. When you pass around a pointer, you make the garbage collection have to follow the pointer through more places to determine whether the block of memory is still in use. That is one reason why it may you may get better performance from copying over passing by pointer. That same thinking would also lend itself to considering `TFPluginInfo` over `*TFPluginInfo` for the value of the map.